### PR TITLE
Feat (optimizer): add new optimizer for spliting zero tensor

### PR DESCRIPTION
**Pull request by li126com**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

A new optimizer which uniformly splits the parameters in the optimizer state into each rank in zero parallelism.
functions implemented：
- [x] dp
- [x] tp
- [x] pp
- [x] fsp and msp
- [x] isp
- [x] zero1.5
- [x] overlap grad
- [x] overlap param
- [x] checkpoint

## Check
trace:
overlap param and grad
image

loss:
dp8:
image

dp2_tp2_pp2_fsp:
image

isp:
image

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: configs/7B_sft.py [hunk 1]
```diff
@@ -100,6 +100,9 @@
     reduce_bucket_size=512 * 1024 * 1024,
     # grad clipping
     clip_grad_norm=1.0,
+    # whether use new optm
+    new_version=False,
```
Comment (huangting4201) [inline configs/7B_sft.py]: 这个字段名太抽象了吧

File: configs/7B_sft.py [hunk 2]
```diff
@@ -100,6 +100,9 @@
...
+    all_gather_size=512 * 1024 * 1024,
```
Comment (huangting4201) [inline configs/7B_sft.py:107]: 这个all_gather_size加上注释？

File: internlm/solver/optimizer/hybrid_zero_optim.py [hunk 3]
```diff
@@ -953,3 +954,23 @@ def reload_zero_fp32_buff(optimizer):
...
+                        else:
+                            padding_param = working_param.data.view(-1)
+
+                        splited_params = padding_param.split(
+                            padding_param.numel() // optimizer._zero_world_size[group_id]
+                        )
+                        splited_params = splited_params[optimizer._zero_local_rank[group_id]]
+                        splited_param_current_rank = splited_params.detach().float()
+
+                    master_param.data.copy_(splited_param_current_rank)
```
Comment (huangting4201) [inline internlm/solver/optimizer/hybrid_zero_optim.py]: 如果处理两种optimizer的话，这个函数移出去放在optimizer/utils.py文件里是不是更合适些

File: internlm/solver/optimizer/hybrid_zero_optim_v2.py [hunk 4]
```diff
@@ -0,0 +1,876 @@
...
+
+                    grad_dtype = flat_grads.dtype
+
+                    if not self._partition_grads:
+                        reduce_grads = torch.zeros(
+                            flat_grads.numel() // self._zero_world_size[group_id],
+                            dtype=grad_dtype,
+                            device=get_current_device(),
+                        )
+                        dist.reduce_scatter_tensor(reduce_grads, flat_grads, group=reduce_group)
```
Comment (li126com) [inline internlm/solver/optimizer/hybrid_zero_optim_v2.py]: zero 1.5 needs to use all reduce

**Review by mwiacx — APPROVED** [review]

File: configs/7B_sft.py (hunk 1 above)
Comment (li126com) [reply configs/7B_sft.py]: 诶，起名字这个我也没想出来什么好的，那该叫什么好呢，split_tensor_optm?

File: configs/7B_sft.py (hunk 2 above)
Comment (li126com) [reply configs/7B_sft.py:107]: done

**Review by sunpengsdu — APPROVED** [review]
